### PR TITLE
Master mrp improvements ajf

### PR DESCRIPTION
**Pull request by ajf-odoo**
Description of the issue/feature this PR addresses:

Current behavior before PR:

Desired behavior after PR is merged:

---
I confirm I have signed the CLA and read the PR guidelines at www.odoo.com/submit-pr

### CODE REVIEW

File: addons/mrp/models/mrp_production.py [hunk 1]
```diff
@@ -38,7 +38,7 @@ def _get_default_date_planned_start(self):
     def _get_default_is_locked(self):
         return not self.user_has_groups('mrp.group_unlocked_by_default')
 
-    name = fields.Char('Reference', copy=False, readonly=True)
+    name = fields.Char('Reference', default='New MO', copy=False, readonly=True)
```
Comment (Whenrow) [inline addons/mrp/models/mrp_production.py:41]: Keep "New" with the model name on top like this 
image

File: addons/mrp/models/mrp_production.py [hunk 2]
```diff
@@ -243,6 +244,9 @@ def _get_default_is_locked(self):
         compute='_compute_show_allocation',
         help='Technical Field used to decide whether the button "Allocation" should be displayed.')
     allow_workorder_dependencies = fields.Boolean('Allow Work Order Dependencies')
+    show_produce = fields.Boolean(compute='_compute_show_produce', help='Technical field to check if produce button can be shown')
```
Comment (Whenrow) [inline addons/mrp/models/mrp_production.py:247]: Move in commit where you remove the immediate wizard

File: addons/mrp/views/mrp_production_views.xml [hunk 3]
```diff
@@ -118,8 +118,8 @@
...
                     <button name="button_scrap" type="object" string="Scrap" attrs="{'invisible': [('state', 'in', ('cancel', 'draft'))]}" data-hotkey="z"/>
-                    <field name="state" widget="statusbar" statusbar_visible="draft,confirmed,progress,done"/>
-                    <button name="action_toggle_is_locked" attrs="{'invisible': ['|', ('show_lock', '=', False), ('is_locked', '=', False)]}" string="Unlock" groups="mrp.group_mrp_manager" type="object" help="Unlock the manufacturing order to adjust what has been consumed or produced." data-hotkey="l"/>
+                    <field name="state" widget="statusbar" statusbar_visible="draft,confirmed,done"/>
+                    <button name="action_toggle_is_locked" attrs="{'invisible': ['|', ('show_lock', '=', False), ('is_locked', '=', False)]}" string="Unlock" groups="mrp.group_mrp_manager" type="object" help="Unlock this manufacturing order to adjust required component consumption quantities during the manufacturing order, or consumed quantities retrospectively, once the production is completed." data-hotkey="l"/>
```
Comment (Whenrow) [inline addons/mrp/views/mrp_production_views.xml:122]: "Unlock to adjust component demand during production, or consumed quantities once the manufacturing order is done."

File: addons/mrp/views/stock_scrap_views.xml [hunk 4]
```diff
@@ -9,6 +9,12 @@
                 <field name="workorder_id" invisible="1"/>
                 <field name="production_id" invisible="1"/>
             </field>
+            <xpath expr="//button[@name='action_validate']" position="attributes">
```
Comment (Whenrow) [inline addons/mrp/views/stock_scrap_views.xml:12]: Make the change in the original view

File: addons/mrp/models/mrp_production.py [hunk 5]
```diff
@@ -52,6 +52,7 @@ def _get_default_is_locked(self):
         'product.product', 'Product',
         domain="""[
             ('type', 'in', ['product', 'consu']),
+            ('id', 'not in', disabled_product_ids),
```
Comment (amoyaux) [inline addons/mrp/models/mrp_production.py:55]: We don't base the domain on dynamic computed field due to performance issue. We aldready removed that kind of domain in the past https://github.com/odoo/odoo/commit/0468e3ccc11042236a3b4bd1c7164a7b780bc481

File: addons/mrp/models/mrp_production.py [hunk 6]
```diff
@@ -1846,19 +1841,44 @@ def button_mark_done(self):
...
+                    if move_line.reserved_uom_qty:
+                        move_line.qty_done = min(move_line.reserved_uom_qty, move_line.move_id.should_consume_qty)
+                    if float_compare(move.quantity_done, move.should_consume_qty, precision_rounding=rounding) >= 0:
+                        break
+                if float_compare(move.product_uom_qty, move.quantity_done, precision_rounding=move.product_uom.rounding) == 1:
+                    if move.has_tracking in ('serial', 'lot'):
+                        missing_lot_id_products += "\n  - %s" % move.product_id.display_name
+            if missing_lot_id_products:
+                error_msg = _('You need to supply Lot/Serial Number for products:') + missing_lot_id_products
+                raise UserError(error_msg)
```
Comment (amoyaux) [inline addons/mrp/models/mrp_production.py:1872]: You could keep that in a separate function. _set_immediate_quantities

File: addons/mrp/models/mrp_production.py [hunk 7]
```diff
@@ -1846,19 +1841,44 @@ def button_mark_done(self):
...
+        if remaining:
```
Comment (amoyaux) [inline addons/mrp/models/mrp_production.py:1873]: if not needed for in empty will skip it anyway

File: addons/mrp/models/mrp_production.py [hunk 8]
```diff
@@ -1846,19 +1841,44 @@ def button_mark_done(self):
         return action
 
     def _pre_button_mark_done(self):
-        productions_to_immediate = self._check_immediate()
-        if productions_to_immediate:
-            return productions_to_immediate._action_generate_immediate_wizard()
-
+        remaining = self.browse()
```
[the remark this answers is not in the document]
Comment (amoyaux) [reply addons/mrp/models/mrp_production.py:1844]: remaining is a bad name. I would keep immediate_productions as before
and keep the same structure as before:

loop to fill production without qty in immediate_productions
immediate_productions._set_immediate_quantities()
loop for production in self - immediate_productions to check consumption warning

**Review by amoyaux — COMMENTED** [review]
Also you could probably launch a runbot OE because it will probably have some other test that fail

**Comment by amoyaux** [comment]
robodoo r+ rebase-ff

**Comment by Whenrow** [comment]
robodoo r+

**Pull request closed without merging**